demos/sample_azure_iot_adu/sample_azure_iot_adu.c

```diff
@@ -835,17 +863,50 @@ static void prvAzureDemoTask( void * pvParameters )
                                                      sampleazureiotPROCESS_LOOP_TIMEOUT_MS );
             configASSERT( xResult == eAzureIoTSuccess );
 
-            /* TODO: REMOVE !xDidDeviceUpdate for NXP once properly implemented */
             if( xProcessUpdateRequest && !xDidDeviceUpdate )
             {
-                xResult = prvDownloadUpdateImageIntoFlash();
-                configASSERT( xResult == eAzureIoTSuccess );
-
-                xResult = prvEnableImageAndResetDevice();
-                configASSERT( xResult == eAzureIoTSuccess );
-
-                xResult = prvSpoofNewVersion();
-                configASSERT( xResult = eAzureIoTSuccess );
+                if( xAzureIoTAduUpdateRequest.xWorkflow.xAction == eAzureIoTADUActionCancel )
+                {
+                    xResult = AzureIoTADUClient_SendAgentState( &xAzureIoTADUClient,
+                                                                &xAzureIoTHubClient,
+                                                                &xADUDeviceProperties,
+                                                                &xAzureIoTAduUpdateRequest,
+                                                                eAzureIoTADUAgentStateIdle,
+                                                                NULL,
+                                                                ucScratchBuffer,
+                                                                sizeof( ucScratchBuffer ),
+                                                                NULL );
+
+                    xProcessUpdateRequest = false;
+                }
+                else
+                {
+                    xResult = prvDownloadUpdateImageIntoFlash( sampleazureiotADU_DOWNLOAD_TIMEOUT );
+                    configASSERT( xResult == eAzureIoTSuccess );
+
+                    if( xAzureIoTAduUpdateRequest.xWorkflow.xAction == eAzureIoTADUActionApplyDownload )
```

Comment:
Is there a scenario where xProcessUpdateRequest is true, xAzureIoTAduUpdateRequest.xWorkflow.xAction != eAzureIoTADUActionCancel and xAzureIoTAduUpdateRequest.xWorkflow.xAction isn't eAzureIoTADUActionApplyDownload?

Reply:
No there's only two values for that enum. That being said, I don't think it's a bad idea to have an explicit check for each and then an else which logs that an unknown value was set. I'll update....good question.

Reply:
Your question also brings up another scenario that I think we need to account for here. We should check for `xProcessUpdateRequest` again in case an invalid ADU payload came in (with some kind of weird `NULL`s or something. I'll add it

Reply:
Agree - it's good future proofing in case a third value comes along

Reply:
> Your question also brings up another scenario that I think we need to account for here. We should check for `xProcessUpdateRequest` again in case an invalid ADU payload came in (with some kind of weird `NULL`s or something. I'll add it

This would be in the scenario that during one of the checks during the timeouts in prvDownloadUpdateImageIntoFlash we get a cancel request? Would that scenario bump us out of this loop though because the assert on 885 `configASSERT( xResult == eAzureIoTSuccess );` wouldn't be true?

Reply:
I think it also might be good to add a comment calling out why we have the last if statement since the code looks like we check `xProcessUpdateRequest && ( xAzureIoTAduUpdateRequest.xWorkflow.xAction == eAzureIoTADUActionApplyDownload )` right next to the other if statements checking these values

Reply:
Yes in the case that the deployment was cancelled in the middle of the download, it breaks the download and returns success. So the assert would not get tripped. Also added extra comments on why we're doing what we're doing.